libraries/persistent_store/src/buffer.rs

```diff
@@ -0,0 +1,668 @@
+            for word in 0..num_words {
+                write!(f, " [{}]", self.word_writes[page * num_words + word])?;
+                for byte in 0..num_bytes {
+                    let index = (page * num_words + word) * num_bytes + byte;
+                    write!(f, "{:02x}", self.storage[index])?;
+                }
+            }
+            writeln!(f)?;
+        }
+        Ok(())
+    }
+}
+
+/// Represents a storage operation.
+///
+/// It is polymorphic over the ownership of the byte slice ot avoid unnecessary copies.
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum BufferOperation<ByteSlice: Borrow<[u8]>> {
+    /// Represents a write operation.
+    Write {
+        /// The storage index at which the write should occur.
+        index: StorageIndex,
+
+        /// The slice that should be written.
+        value: ByteSlice,
+    },
+
+    /// Represents an erase operation.
+    Erase {
+        /// The page that should be erased.
+        page: usize,
+    },
+}
+
+/// Represents a storage operation owning its byte slices.
+type OwnedBufferOperation = BufferOperation<Box<[u8]>>;
+
+/// Represents a storage operation sharing its byte slices.
+type SharedBufferOperation<'a> = BufferOperation<&'a [u8]>;
+
+impl<'a> SharedBufferOperation<'a> {
+    fn to_owned(&self) -> OwnedBufferOperation {
+        match *self {
+            BufferOperation::Write { index, value } => BufferOperation::Write {
+                index,
+                value: value.to_vec().into_boxed_slice(),
+            },
+            BufferOperation::Erase { page } => BufferOperation::Erase { page },
+        }
+    }
+}
+
+/// Controls when an operation is interrupted.
+///
+/// This can be used to simulate power-offs while the device is writing to the storage or erasing a
+/// page in the storage.
+#[derive(Clone)]
+enum Interruption {
+    /// Mutable operations have normal behavior.
+    Ready,
+
+    /// If the delay is positive, mutable operations decrement it. If the count is zero, mutable
+    /// operations fail and are saved.
+    Armed { delay: usize },
+
+    /// Mutable operations panic.
+    Saved { operation: OwnedBufferOperation },
+}
+
+impl Interruption {
+    /// Arms an interruption for a given delay.
+    ///
+    /// # Panics
+    ///
+    /// Panics if an interruption is already armed.
+    fn arm(&mut self, delay: usize) {
+        match self {
+            Interruption::Ready => *self = Interruption::Armed { delay },
+            _ => panic!(),
+        }
+    }
+
+    /// Disarms an interruption.
+    ///
+    /// Returns the interrupted operation if any, otherwise the remaining delay.
+    ///
+    /// # Panics
+    ///
+    /// Panics if an interruption was not armed.
+    fn get(&mut self) -> Result<OwnedBufferOperation, usize> {
+        let mut interruption = Interruption::Ready;
+        core::mem::swap(self, &mut interruption);
+        match interruption {
+            Interruption::Armed { delay } => Err(delay),
```

Comment:
Yes, `get` always disarms. It just returns something different depending on whether it triggered or not.

For more context, this is because in fuzzing the moment when an operation is interrupted is "random", so it might happen that we don't reach that point before the next "full analysis" (checking invariants, etc.) after which we want to resume everything as normal under full control of the fuzzing driver.

Reply:
Oh, I missed the first documentation line. Resolved.